### PR TITLE
Add docker publishing scripts and pipeline

### PR DESCRIPTION
**Pull request by navarone-feekery**
### Related to https://github.com/elastic/crawler/issues/41

Add scripts to publish docker image to Elastic's docker repository.

Creation of a pipeline for scheduled releases of Docker images for Elastic Connectors. This builds, tests, and pushes Docker images for `arm64`, `amd64`, and a multi-architecture image containing both.

This adds a pipeline and scripts that:
* Builds docker images for each platform (and runs a quick test on each to verify they look correct)
* Pushes the images to the registry
* Builds a multi-arch image from the individual pushed images

### CODE REVIEW

File: catalog-info.yaml
Comment (seanstory) [inline catalog-info.yaml:46]:
```suggestion
  name: "crawler-docker-build-publish"
```

File: .buildkite/scripts/publish_docker.sh
Comment (seanstory) [inline .buildkite/scripts/publish_docker.sh:25]: These need to be replaced with the vault paths and field names for the docker user/password secrets that are specific to Crawler.

File: catalog-info.yaml
Comment (seanstory) [inline catalog-info.yaml:50]: :sus: this 404's, and IDK why. I would have expected to be able to run this pipeline off of this PR branch to test it.

We may need to first merge this `catalog-info.yml` change along with an empty pipeline, and then _iterate_ on that pipeline in a subsequent PR.

File: .buildkite/scripts/publish/publish-common.sh
Comment (seanstory) [inline .buildkite/scripts/publish/publish-common.sh:24]: There's been a thread going with leads about what the right repository/namespace is for our stuff going forward. I've bumped the thread, but worth pinging Diana/Dana too. But I don't think `/enterprise-search` is going to be the right place for this, and we might as well start in the right place 🤞

**Review by seanstory — COMMENTED** [review]
Few nits, but mostly looks good! I suggest you get a cursory review from @elastic/search-productivity-team too

File: .buildkite/scripts/publish_docker.sh
Comment (navarone-feekery) [reply .buildkite/scripts/publish_docker.sh:25]: Idk how I missed this one 🫠

File: .buildkite/scripts/publish/publish-common.sh
Comment (navarone-feekery) [reply .buildkite/scripts/publish/publish-common.sh:24]: Once a decision is made about the future namespace, do you have an idea of how long it would take to set up? Just thinking about how long this PR may have to wait for a decision + implementation.

File: catalog-info.yaml
Comment (navarone-feekery) [reply catalog-info.yaml:50]: I just assumed it would be automagically created once merging this PR.
Here's a PR for that: https://github.com/elastic/crawler/pull/104

Comment (seanstory) [reply catalog-info.yaml:50]: I'd thought so too. 🤷

File: .buildkite/scripts/publish/publish-common.sh
Comment (seanstory) [reply .buildkite/scripts/publish/publish-common.sh:24]: I don't think setup is required - I think you'll just be able to push wherever you tell it to go. But we'll find out!
I'm hoping we can hear back from leads by tomorrow/Monday

Comment (acrewdson) [reply .buildkite/scripts/publish/publish-common.sh:24]: related to this, there's also a slack thread from a few weeks ago https://elastic.slack.com/archives/C7UQBTBD0/p1722424618694589

and a related github issue: https://github.com/elastic/search-team/issues/7481

File: .buildkite/scripts/publish_docker.sh
Comment (acrewdson) [inline .buildkite/scripts/publish_docker.sh:30]: this should be fine, but a nice future improvement could be to consider using the [vault-secrets](https://github.com/elastic/vault-secrets-buildkite-plugin) buildkite plugin instead, as described in [the docs here](https://docs.elastic.dev/ci/using-secrets#use-in-ci-pipelines)

File: .buildkite/scripts/publish/publish-common.sh
Comment (acrewdson) [inline .buildkite/scripts/publish/publish-common.sh:19]: how does `USE_SNAPSHOT` get set in practice?

File: .buildkite/release-pipeline.yml
Comment (acrewdson) [inline .buildkite/release-pipeline.yml:1]: this may be out of scope, but to me it looks like the images we are going to publish here will contain hundreds of CVEs

it could be nice to see if we can invest the effort up-front to start by publishing [Wolfi-based images](https://docs.elastic.dev/wolfi), instead

File: .buildkite/scripts/publish/build-multiarch-docker.sh
Comment (acrewdson) [inline .buildkite/scripts/publish/build-multiarch-docker.sh:1]: should this script be renamed to include a reference to `push`, since it isn't just building, but is also pushing?

File: Makefile
Comment (acrewdson) [inline Makefile:37]: ~doesn't `-SNAPSHOT` get added via the `USE_SNAPSHOT` stuff in `publish-common.sh`, so would be redundant here?~

actually maybe a better question is: why are we adding this Makefile target (and the one elow) if we do our image buliding and pushing via scripts in `.buildkite/scripts/publish`?

File: .buildkite/scripts/publish/publish-common.sh
Comment (acrewdson) [inline .buildkite/scripts/publish/publish-common.sh:1]: could be worth running shellcheck on this and other bash scripts being added, I think it'll flag some things to fix

File: .buildkite/scripts/publish_docker.sh
Comment (acrewdson) [inline .buildkite/scripts/publish_docker.sh:1]: is this script being used anywhere?

File: .buildkite/scripts/publish/publish-common.sh
Comment (acrewdson) [inline .buildkite/scripts/publish/publish-common.sh:28]: I don't think we can read directly from this Vault within a Buildkite job without doing some extra work, _eg_ the 'transitive auth' approach I used here https://github.com/elastic/swiftype-ci-base/blob/master/.buildkite/hooks/pre-command

another option would be to duplicate the secrets in CI vault, in which case we could then use [this plugin](https://github.com/elastic/vault-secrets-buildkite-plugin) to make secret injection a bit nicer

File: .buildkite/scripts/publish/publish-common.sh
Comment (acrewdson) [inline .buildkite/scripts/publish/publish-common.sh:29]: nit: `VAULT_PATH` would be more idiomatic

File: .buildkite/scripts/publish/test-docker.sh
Comment (acrewdson) [inline .buildkite/scripts/publish/test-docker.sh:73]:
```suggestion
  # ensure JRuby 9.4.* is installed
  - name: "JRuby Installation 9.4.*"
```

File: .buildkite/scripts/publish/publish-common.sh
Comment (navarone-feekery) [reply .buildkite/scripts/publish/publish-common.sh:19]: It can be set in the env when running from buildkite. Here's a screenshot of the UI for Connectors:
Screenshot 2024-08-23 at 10 22 15

File: .buildkite/release-pipeline.yml
Comment (navarone-feekery) [reply .buildkite/release-pipeline.yml:1]: I think this sounds like a good idea. I'll try and switch this to Wolfi today.

File: Makefile
Comment (navarone-feekery) [reply Makefile:37]: I included this so we'd have a manual way to publish from local, but thinking more on it, I don't think we'll ever do that. It can be deleted.

File: .buildkite/scripts/publish/publish-common.sh
Comment (navarone-feekery) [reply .buildkite/scripts/publish/publish-common.sh:1]: I didn't know shellcheck was a thing, thanks for the suggestion! I've fixed everything it flagged.

File: .buildkite/scripts/publish_docker.sh
Comment (navarone-feekery) [reply .buildkite/scripts/publish_docker.sh]: It was part of the publish locally workflow. I've removed it along with the makefile commands.

File: .buildkite/scripts/publish/publish-common.sh
Comment (navarone-feekery) [reply .buildkite/scripts/publish/publish-common.sh]: Unfortunately I don't have access to the example you've linked.
I think moving the credentials to CI vault makes sense. I'll do that (and look into using the plugin)

File: .buildkite/release-pipeline.yml
Comment (navarone-feekery) [reply .buildkite/release-pipeline.yml:1]: I had an attempt at this today but I'm running into various issues. I made a draft PR: https://github.com/elastic/crawler/pull/106
It doesn't work yet, but we can iterate on it.

File: .buildkite/scripts/publish/publish-common.sh
Comment (navarone-feekery) [reply .buildkite/scripts/publish/publish-common.sh]: I've moved it to the CI vault and updated everything.
I haven't tried to use the plugin. I can see it's cleaner but because everything is already written, I'd prefer not to rewrite it (also it's identical to Connectors which makes maintaining easier for the team).

File: .buildkite/scripts/publish/publish-common.sh
Comment (navarone-feekery) [reply .buildkite/scripts/publish/publish-common.sh]: FYI I've addressed all of the review comments, and I think docker repo namespace is the only blocker now for this PR.

Comment (seanstory) [reply .buildkite/scripts/publish/publish-common.sh]: Latest I've seen from Paul is to go with `docker.elastic.co/integrations/crawler`. Let's go with that. Worst case scenario, we publish a few images there then have to pivot later.

File: .buildkite/release-pipeline.yml
Comment (acrewdson) [reply .buildkite/release-pipeline.yml:1]: nice! I added a thought over there https://github.com/elastic/crawler/pull/106#pullrequestreview-2257621613

File: .buildkite/scripts/publish/publish-common.sh
Comment (acrewdson) [reply .buildkite/scripts/publish/publish-common.sh:23]: ah, gotcha, wasn't sure if that was how we were expecting to set the option

in that case, could it be a nicer user experience to use a formal input step, to make the existence of this snapshot option more obvious to people kicking off builds? https://buildkite.com/docs/pipelines/input-step

File: .buildkite/scripts/publish/publish-common.sh
Comment (acrewdson) [reply .buildkite/scripts/publish/publish-common.sh]: > FYI I've addressed all of the review comments, and I think docker repo namespace is the only blocker now for this PR.

thanks for the fixes, this is looking great to me 🎉 

I'll leave it to you all to decide whether we'd want to consider switching to Wolfi for the base image as a blocker (main rationale being: let's not start by publishing a new image with hundreds of CVEs). I'd guess it depends on how soon we need an official crawler image available.

Comment (acrewdson) [reply .buildkite/scripts/publish/publish-common.sh]: update to the above, it's looking to me like a wolfi-based image may be easily within reach: https://github.com/elastic/crawler/pull/106#issuecomment-2307878798

I was able to build a wolfi-based image locally and start the crawler:
```
$ docker run -it crawler-wolfi:latest
Commands:
  crawler crawl CRAWL_CONFIG                   # Run a crawl of the site
  crawler validate CRAWL_CONFIG                # Validate crawler configuration
  crawler version                              # Print version
```

Comment (navarone-feekery) [reply .buildkite/scripts/publish/publish-common.sh]: > Latest I've seen from Paul is to go with docker.elastic.co/integrations/crawler. Let's go with that. Worst case scenario, we publish a few images there then have to pivot later.

@seanstory I've updated the base docker tag to be this value.

Comment (navarone-feekery) [reply .buildkite/scripts/publish/publish-common.sh]: @acrewdson thanks heaps for iterating on the wolfi image. I'd like to have it from day one, but I'd prefer to merge it in from the other PR. Merging this PR won't produce an image yet so I think it's okay to do them separately.

File: .buildkite/scripts/publish/publish-common.sh
Comment (navarone-feekery) [reply .buildkite/scripts/publish/publish-common.sh]: I didn't know this was a thing 🤯 
I've added a step for this in this commit: https://github.com/elastic/crawler/pull/103/commits/326e36ef92b5523206f9d1655696475ab9b8e90a

Comment (acrewdson) [reply .buildkite/scripts/publish/publish-common.sh]: nice!

File: .buildkite/scripts/publish/publish-common.sh
Comment (acrewdson) [reply .buildkite/scripts/publish/publish-common.sh]: > I'd like to have it from day one, but I'd prefer to merge it in from the other PR. Merging this PR won't produce an image yet so I think it's okay to do them separately.

makes sense 👍

**Review by acrewdson — APPROVED** [review]

File: .buildkite/scripts/publish/publish-common.sh
Comment (navarone-feekery) [reply .buildkite/scripts/publish/publish-common.sh]: > Suggestion: could search-integrations be more future-proof than integrations? Please discard this comment if this was already extensively discussed.

@oli-g this was discussed among the leads, but I don't have a thread available to reference. @seanstory is the thread where this was discussed shareable? If not, do you think the current naming is fine?

**Review by acrewdson — DISMISSED** [review]

File: .buildkite/scripts/publish/publish-common.sh
Comment (navarone-feekery) [reply .buildkite/scripts/publish/publish-common.sh]: Confirmed that `integrations` is the official answer.

File: .buildkite/publish/publish-common.sh
Comment (acrewdson) [inline .buildkite/publish/publish-common.sh:30]: @navarone-feekery now that we have a working Wolfi-based Dockerfile, do you think we could try using it here, at this point?

Comment (navarone-feekery) [reply .buildkite/publish/publish-common.sh:30]: I also think this PR should switch to wolfi, I'll work on that today

Comment (navarone-feekery) [reply .buildkite/publish/publish-common.sh:30]: It built fine with the wolfi image 🎉 
https://buildkite.com/elastic/crawler-docker-build-publish/builds/14

**Comment by navarone-feekery** [comment]
I was able to build and publish the artifact: https://buildkite.com/elastic/crawler-docker-build-publish/builds/14

**Review by oli-g — APPROVED** [review]

**Comment by seanstory** [comment]
🎉 Yay!

@navarone-feekery I'm not seeing the images at https://container-library.elastic.co/r/integrations/crawler or https://www.docker.elastic.co/r/integrations/crawler. Though if I log in at https://container-library.elastic.co and search for `crawler` it pre-populates `integrations/crawler` (which then 404s if clicked on).

Any idea what's going on?

**Comment by seanstory** [comment]
Ok, I think I found breadcrumbs as this felt familiar: https://github.com/elastic/infra/issues/40122
I'll start a slack thread and loop you in.

**Pull request merged**
